src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -103,27 +104,31 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
         Consumes the rate limits returned by `check_write_limits`.
         """
         if exc_type is None:
-            self._writes_limiter._get_rate_limiter(self._use_case_id, self._namespace).use_quotas(
-                self._requests, self._grants, self._timestamp
-            )
+            self._writes_limiter._get_rate_limiter(
+                self._use_case_id, self._namespace, self._db_backend
+            ).use_quotas(self._requests, self._grants, self._timestamp)
 
 
 class WritesLimiter:
     def __init__(self) -> None:
         self.rate_limiters: MutableMapping[str, RedisSlidingWindowRateLimiter] = {}
 
     def _get_rate_limiter(
-        self, use_case_id: UseCaseKey, namespace: str
+        self, use_case_id: UseCaseKey, namespace: str, db_backend: IndexerStorage
     ) -> RedisSlidingWindowRateLimiter:
         if namespace not in self.rate_limiters:
-            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            options = get_ingest_config(use_case_id, db_backend).writes_limiter_cluster_options
             self.rate_limiters[namespace] = RedisSlidingWindowRateLimiter(**options)
 
         return self.rate_limiters[namespace]
 
     @metrics.wraps("sentry_metrics.indexer.check_write_limits")
     def check_write_limits(
-        self, use_case_id: UseCaseKey, namespace: str, keys: KeyCollection
+        self,
+        use_case_id: UseCaseKey,
+        namespace: str,
+        db_backend: IndexerStorage,
+        keys: KeyCollection,
```

Comment:
I thought the idea of the namespace was to avoid the rate limtiers class to know about the DB backend. It does not seem right that a rate limiter cares whether it is rate limiting spanner or postgres. The rate limiter should work in the same way independently.
What is the reason to introduce the dependency on the DB backend ?

Reply:
So the problem I ran into was that we get the config in the ratelimiter https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L119 so that we can get the redis options, if the ratelimiter doens't already exist. 

I feel like it could make more sense to initialize the writes_limiter with the namespace and options, or maybe have like an add_namespace method on the `WritesLimiter` so that when we call `check_write_limits` we aren't passing the `db_backend`.

Reply:
I see. 
I think the issue is here 
https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/indexer/ratelimiters.py#L185

This class is meant to exist in one instance only.
Obviously now we have multiple instances at the same time so having one static instance does not work anymore and we need to be able to return the right rate limiter when we try to write.

> I feel like it could make more sense to initialize the writes_limiter with the namespace and options

I think this is right, a good idea would be to move the options in the initialization of the `WritesLimiter` class and then have a factory that returns the right instance based on the config.